protocols/randhound/proof.go

```diff
@@ -0,0 +1,298 @@
+package randhound
+
+import (
+	"errors"
+
+	"github.com/dedis/cothority/crypto"
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/poly"
+	"github.com/dedis/crypto/random"
+)
+
+// Package (XXX: name) provides functionality to show equality of discrete
+// logarithms (dlog) through non-interactive zero-knowledge (NIZK) proofs.
+
+// Proof resembles a NIZK dlog-equality proof. Allows to handle multiple proofs.
+type Proof struct {
+	suite abstract.Suite
+	base  []ProofBase
+	core  []ProofCore
+}
+
+// ProofBase contains the base points against which the core proof is created.
+type ProofBase struct {
+	g abstract.Point
+	h abstract.Point
+}
+
+// ProofCore contains the core elements of the NIZK dlog-equality proof.
+type ProofCore struct {
+	C  abstract.Scalar // challenge
+	R  abstract.Scalar // response
+	VG abstract.Point  // public commitment with respect to base point G
+	VH abstract.Point  // public commitment with respect to base point H
+}
+
+// NewProof creates a new NIZK dlog-equality proof.
+func NewProof(suite abstract.Suite, g []abstract.Point, h []abstract.Point, core []ProofCore) (*Proof, error) {
+
+	if len(g) != len(h) {
+		return nil, errors.New("Received non-matching number of points")
+	}
+
+	n := len(g)
+	base := make([]ProofBase, n)
+	for i := range base {
+		base[i] = ProofBase{g: g[i], h: h[i]}
+	}
+
+	return &Proof{suite: suite, base: base, core: core}, nil
+}
+
+// Setup initializes the proof by randomly selecting a commitment v,
+// determining the challenge c = H(xG,xH,vG,vH), and the response r = v - cx.
+func (p *Proof) Setup(scalar ...abstract.Scalar) ([]abstract.Point, []abstract.Point, []ProofCore, error) {
+
+	if len(scalar) != len(p.base) {
+		return nil, nil, nil, errors.New("Received unexpected number of scalars")
+	}
+
+	n := len(scalar)
+	p.core = make([]ProofCore, n)
+	xG := make([]abstract.Point, n)
+	xH := make([]abstract.Point, n)
+	for i, x := range scalar {
+
+		xG[i] = p.suite.Point().Mul(p.base[i].g, x)
+		xH[i] = p.suite.Point().Mul(p.base[i].h, x)
+
+		// Commitment
+		v := p.suite.Scalar().Pick(random.Stream)
+		vG := p.suite.Point().Mul(p.base[i].g, v)
+		vH := p.suite.Point().Mul(p.base[i].h, v)
+
+		// Challenge
+		cb, err := crypto.HashArgsSuite(p.suite, xG[i], xH[i], vG, vH)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		c := p.suite.Scalar().Pick(p.suite.Cipher(cb))
+
+		// Response
+		r := p.suite.Scalar()
+		r.Mul(x, c).Sub(v, r)
+
+		p.core[i] = ProofCore{c, r, vG, vH}
+	}
+
+	return xG, xH, p.core, nil
```

Comment:
> Would it be better coding style to have a "getter"?

I think this depends on your preferences. As `p.core` is an array, the returned object (in any case) can be manipulated. So you could as well just make `core` a public/exported field:
See here: https://play.golang.org/p/pkZHSVzSgk